h2/src/main/org/h2/command/dml/Insert.java

```diff
@@ -204,15 +205,34 @@ private int insertRows() {
                 query.query(0, this);
             } else {
                 ResultInterface rows = query.query(0);
+                int updatedRows = 0;
                 while (rows.next()) {
                     generatedKeys.nextRow();
                     Value[] r = rows.currentRow();
-                    Row newRow = addRowImpl(r);
-                    if (newRow != null) {
-                        generatedKeys.confirmRow(newRow);
+                    try {
+                        Expression[] exp = new Expression[columns.length];
+                        for (int j = 0, len = columns.length; j < len; j++) {
+                            exp[j] = ValueExpression.get(r[j]);
+                        }
+                        addRow(exp);
```

Comment:
I don't think that this is a good solution. Number of rows can be very large. It's better to pass current row to `handleOnDuplicate()` as additional parameter.

Reply:
Done